tutorial_src/lea/prog/leaingen.cpp

```diff
@@ -0,0 +1,172 @@
+        string name = id + to_string(no) + *letter + ".in";
+        cerr << "Printing to file: " << name << "\n";
+        ofstream outFile(name);
+        outFile << n << "\n";
+        for (auto edge : edges) {
+            outFile << edge.first << " " << edge.second << "\n";
+        }
+        outFile.close();
+        (*letter)++;
+    }
+};
+
+// Shuffles vertices indexes and edges order
+void shuffle(Test *test) {
+    RG.randomShuffle(test->edges.begin(), test->edges.end());
+    for (int i = 0; i < test->n - 1; i++) {
+        if (randomUIntInRange(0, 1)) {
+            swap(test->edges[i].first, test->edges[i].second);
+        }
+    }
+    
+    vector<int> mapping;
+    for (int i = 0; i <= test->n; i++) {
+        mapping.push_back(i);
+    }
+    RG.randomShuffle(mapping.begin() + 1, mapping.end());
+    
+    for (int i = 0; i < test->n - 1; i++) {
+        test->edges[i] = make_pair(mapping[test->edges[i].first], mapping[test->edges[i].second]);
+    }
+}
+
+// Shuffles nodes indexes until vertex 1 has degree equal to one or not.
+// Works only when about half of vertices are leaves,
+// otherwise shuffling might take really long.
+Test forceRootDegree(Test test, bool forceToOne) {
+    while (true) {
+        int rootDegree = 0;
+        for (int i = 0; i < test.n - 1; i++) {
+            if (test.edges[i].first == 1 || test.edges[i].second == 1) {
+                rootDegree++;
+            }
+        }
+        if (rootDegree == 1 && forceToOne)
+            break;
+        if (rootDegree > 1 && !forceToOne)
+            break;
+        shuffle(&test);
+    }
+    return test;
+}
+
+// Path and star tests have a probability close to 1
+// that root has/has not the degree we want.
+// We trust it, worst case scenario inwer will inform us we are wrong.
+
+Test path(int _n, bool rootDegreeIsOne) {
+    Test result = Test(_n);
+    for (int i = 2; i <= result.n; i++) {
+        result.edges.push_back({i-1, i});
+    }
+
+    if (!rootDegreeIsOne)
+        shuffle(&result);
+
+    return result;
+}
+
+Test star(int _n, bool rootDegreeIsOne) {
+    Test result = Test(_n);
+    for (int i = 2; i <= result.n; i++) {
+        result.edges.push_back({1, i});
+    }
+
+    if (rootDegreeIsOne)
+        shuffle(&result);
+
+    return result;
+}
+
+// Binary tree and random tests have a probability close to 0.5
+// that the root has/has not the degree we want.
+// Therefor we can safely shuffle indexes till we get the value we want.
+
+Test binaryTree(int _n) {
+    Test result = Test(_n);
+    for (int i = 2; i <= result.n; i++) {
+        result.edges.push_back({i/2, i});
+    }
+
+    shuffle(&result);
+    return result;
+}
+
+Test random(int _n) {
+    Test result = Test(_n);
+    for (int i = 2; i <= result.n; i++) {
+        result.edges.push_back(make_pair(randomUIntInRange(1, i-1), i));
+    }
+
+    shuffle(&result);
+    return result;
+}
+
+int main() {
+    
+    { // Test group 1.
```

Comment:
Tutaj przydałoby się opisać, co to są za grupy. Duża jest ta generatorka i trudno mi się zorientować, co ona robi.